lib/spack/spack/test/error_messages.py

```diff
@@ -0,0 +1,554 @@
+    with open(repo_path / "__init__.py", "w", encoding="utf-8"):
+        pass
+    repo_yaml = os.path.join(repo_path, "repo.yaml")
+    with open(str(repo_yaml), "w", encoding="utf-8") as f:
+        f.write(
+            f"""\
+repo:
+  namespace: {repo_name}
+  api: v2.1
+"""
+        )
+
+    _repo_name_id += 1
+
+    packages_dir = repo_path / "packages"
+    os.mkdir(packages_dir)
+    with open(packages_dir / "__init__.py", "w", encoding="utf-8"):
+        pass
+    for pkg_name, pkg_str in pkg_name_content_tuples:
+        pkg_dir = packages_dir / pkg_name
+        os.mkdir(pkg_dir)
+        pkg_file = pkg_dir / "package.py"
+        with open(str(pkg_file), "w", encoding="utf-8") as f:
+            f.write(pkg_str)
+
+    repo_cache = spack.util.file_cache.FileCache(str(tmp_path / "cache"))
+    return spack.repo.Repo(str(repo_path), cache=repo_cache)
+
+
+@pytest.fixture
+def _create_test_repo(tmp_path, mutable_config):
+    yield create_test_repo(tmp_path, all_pkgs)
+
+
+@pytest.fixture
+def test_repo(_create_test_repo, monkeypatch, mock_stage):
+    with spack.repo.use_repositories(_create_test_repo) as mock_repo_path:
+        yield mock_repo_path
+
+
+@contextmanager
+def expect_failure_and_print(should_mention=None):
+    got_an_error_as_expected = False
+    err_msg = None
+    try:
+        yield
+    except spack.error.UnsatisfiableSpecError as e:
+        got_an_error_as_expected = True
+        err_msg = str(e)
+    if not got_an_error_as_expected:
+        raise ValueError("A failure was supposed to occur in this context manager")
+    elif not err_msg:
+        raise ValueError("No error message for failed concretization")
+    print(err_msg)
+    check_error(err_msg, should_mention)
+
+
+def check_error(msg, should_mention: Optional[Iterable] = None):
+    excludes = [
+        "failed to concretize .* for the following reasons:",
+        "Cannot satisfy .*",
+        "required because .* requested explicitly",
+        "cannot satisfy a requirement for package .*",
+    ]
+    lines = msg.split("\n")
+    should_mention = set(should_mention) if should_mention else set()
+    should_mention_hits = set()
+    remaining = []
+    for line in lines:
+        if any(re.search(p, line) for p in excludes):
+            continue
+        for p in should_mention:
+            if re.search(p, line):
+                should_mention_hits.add(p)
+        remaining.append(line)
+    if not remaining:
+        raise ValueError("The error message contains only generic statements")
+    should_mention_misses = should_mention - should_mention_hits
+    if should_mention_misses:
+        raise ValueError(f"The error message did not contain: {sorted(should_mention_misses)}")
+
+
+# Error message is good
+def test_diamond_with_pkg_conflict1(concretize_scope, test_repo):
+    concretize_one("x2")
+    concretize_one("x3")
+    concretize_one("x4")
+
+    with expect_failure_and_print():
+        concretize_one("x1")
```

Comment:
We should make these tests check more about the error message to make sure it stays good -- there's nothing here that actually tests for a good error message

Reply:
I oriented the checking toward making sure the error message is not bad, vs. checking that it is good.

e..g `test_diamond_with_pkg_conflict1` full error message is:

```
     1. Cannot satisfy 'x4@4.0' 4(4.1)
        required because x3 depends on x4@4.0 
     2. Cannot satisfy 'x4@4.0' 2(4.1)
     3. Cannot satisfy 'x4@4.1' and 'x4@4.0'
        required because x2 depends on x4@4.1 
        required because x3 depends on x4@4.0 
```

the methods `expect_failure_and_print`/`check_error` filter out all the generic error message content and make sure there is something left. Are you thinking we should in fact just compare against that string (maybe minus ordering)? I thought that might make the tests fragile.